src/intersection-observer-polyfill.js

```diff
@@ -268,3 +353,84 @@ export function getThresholdSlot(sortedThreshold, ratio) {
   }
   return endIdx;
 }
+
+/**
+ * Helper function to calculate the intersectionRect and intersection ration.
+ * @param {!./layout-rect.LayoutRectDef} element element's rect
+ * @param {?./layout-rect.LayoutRectDef} owner element's owner rect
+ * @param {!./layout-rect.LayoutRectDef} hostViewport hostViewport's rect
+ * @param {./layout-rect.LayoutRectDef=} opt_iframe iframe container rect
+ * @return {!./layout-rect.LayoutRectDef}
+ */
+function calculateIntersectionRect(element, owner, hostViewport, opt_iframe) {
+  dev().assert(element.width >= 0 && element.height >= 0,
+      'Negative dimensions in element.');
+  // If opt_iframe is provided, all LayoutRect has position relative to
+  // the iframe.
+  // If opt_iframe is not provided, all LayoutRect has position relative to
+  // the host document.
+
+  // Building an IntersectionObserverEntry.
+  let intersectionRect = element;
+  const nonIntersectRect = layoutRectLtwh(0, 0, 0, 0);
+  // element intersects with its owner.
+  if (owner) {
+    intersectionRect = rectIntersection(owner, element) || nonIntersectRect;
+  }
+  // element intersects with container iframe.
+  if (opt_iframe) {
+    intersectionRect = rectIntersection(opt_iframe, intersectionRect) ||
+        nonIntersectRect;
+  }
+  // element intersects with hostViewport
+  intersectionRect = rectIntersection(hostViewport, intersectionRect) ||
+      nonIntersectRect;
+
+  return intersectionRect;
+}
+
+/**
+ * Helper function to calculate the IntersectionObserver change entry.
+ * @param {!./layout-rect.LayoutRectDef} element element's rect
+ * @param {!./layout-rect.LayoutRectDef} hostViewport hostViewport's rect
+ * @param {!./layout-rect.LayoutRectDef} intersection
+ * @param {number} ratio
+ * @param {./layout-rect.LayoutRectDef=} opt_iframe iframe container rect
+ * @return {!IntersectionObserverEntry}}
+ */
+function calculateChangeEntry(
+    element, hostViewport, intersection, ratio, opt_iframe) {
```

Comment:
As `hostViewport` is not necessary without `opt_iframe`, we could combine the two:

- If `opt_iframe`, pass in the `hostViewport`
- If not, pass in `null` as `hostViewport`